flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -262,7 +262,7 @@ protected StreamTask(
 		this.accumulatorMap = getEnvironment().getAccumulatorRegistry().getUserMap();
 		this.recordWriter = createRecordWriterDelegate(configuration, environment);
 		this.executionDecorator = Preconditions.checkNotNull(executionDecorator);
-		this.mailboxProcessor = new MailboxProcessor(this::processInput);
+		this.mailboxProcessor = new MailboxProcessor(this::processInput, this.executionDecorator);
```

Comment:
With this, checkpointing and processing time timers actions are acquiring checkpoint lock twice. I think `invokeProcessingTimeCallback` and `performCheckpoint` should stop using the decorator explicitly? (`triggerCheckpointOnBarrier` should probably still use it)

Reply:
Inlining `invokeProcessingTimeCallback` and removing sync inside.
As of `performCheckpoint()`, I don't think it's safe, because it's accessible via public methods. I think we should check the difference in benchmarks first.

Reply:
I think `invokeProcessingTimeCallback` was not addressed.

Performance for `performCheckpoint` will not matter, however double acquiring the lock there I think is some kind of code smell?

`performCheckpoint` is used from two contexts:

1. `triggerCheckpointOnBarrier`
2. `triggerCheckpointAsync`

The latter is used as a mailbox action so whole code is now being executed under the checkpoint lock anyway - for this case `performCheckpoint` can safely drop the lock altogether.

For the first one, it's being executed from within the default action and it's also executing `runSynchronousSavepointMailboxLoop`, which maybe shouldn't be done under the `checkpointLock`. 

Maybe it would be cleaner If we had two versions of `performCheckpoint`, one safe acquiring the `checkpointLock` on it's own (used for `triggerCheckpointOnBarrier`). Second unsafe (with `Unsafe` suffix), that doesn't acquire the lock on its own (used in `triggerCheckpointAsync`)?

Reply:
> I think invokeProcessingTimeCallback was not addressed.

Thanks, addressed earlier.


Isn't extra complexity introduced by having two methods doing essentially the same thing is a smell?